src/models/link/metadata.rs

```diff
@@ -0,0 +1,180 @@
+use std::fs::File;
+use std::io::BufReader;
+
+use crate::crypto::{self, PrivateKey};
+use crate::interchange::DataInterchange;
+use crate::Result;
+
+use crate::models::{SignedMetadata, Metadata, Link, VirtualTargetPath, TargetDescription};
+
+/// Helper to construct `LinkMetadata`.
+pub struct LinkMetadataBuilder {
+  name: String,
+  materials: BTreeMap<VirtualTargetPath, TargetDescription>,
+  products: BTreeMap<VirtualTargetPath, TargetDescription>,
+  env: BTreeMap<String, String>,
+  byproducts: BTreeMap<String, String>,
+}
+
+impl LinkMetadataBuilder {
+
+  // This should definitely be improved
+  pub fn new() -> Self {
+      LinkMetadataBuilder {
+          name: String::new(),
+          materials: BTreeMap::new(),
+          products: BTreeMap::new(),
+          env: BTreeMap::new(),
+          byproducts: BTreeMap::new()
+      }
+  }
+
+  /// Set the name number for this link
+  pub fn name(mut self, name: String) -> Self {
+      self.name= name;
+      self
+  }
+
+  /// Set the materials for this metadata
+  pub fn materials(mut self, materials: BTreeMap<VirtualTargetPath, TargetDescription>) -> Self {
+      self.materials = materials;
+      self
+  }
+
+  /// Set the products for this metadata
+  pub fn products(mut self, products: BTreeMap<VirtualTargetPath, TargetDescription>) -> Self {
+      self.products = products;
+      self
+  }
+
+  pub fn add_material(mut self, material_path: VirtualTargetPath) -> Self {
+      let file = File::open(material_path.to_string()).unwrap();
+      let mut reader = BufReader::new(file);
+      let (_length, hashes) = crypto::calculate_hashes(&mut reader,
+              &[crypto::HashAlgorithm::Sha256]).unwrap();
+      self.materials.insert(material_path, hashes);
+      self
+  }
+
+  pub fn add_product(mut self, material_path: VirtualTargetPath) -> Self {
+      let file = File::open(material_path.to_string()).unwrap();
+      let mut reader = BufReader::new(file);
+      let (_length, hashes) = crypto::calculate_hashes(&mut reader,
+              &[crypto::HashAlgorithm::Sha256]).unwrap();
+      self.products.insert(material_path, hashes);
+      self
+  }
+
+  /// Set the products for this metadata
+  pub fn env(mut self, env: BTreeMap<String, String>) -> Self {
+      self.env = env;
+      self
+  }
+
+  /// Set the products for this metadata
+  pub fn byproducts(mut self, byproducts: BTreeMap<String, String>) -> Self {
+      self.byproducts = byproducts;
+      self
+  }
+
+  pub fn build(self) -> Result<LinkMetadata> {
+      LinkMetadata::new(self.name, self.materials, self.products,
+          self.env, self.byproducts)
+  }
+
+    /// Construct a new `SignedMetadata<D, LinkMetadata>`.
+  pub fn signed<D>(self, private_key: &PrivateKey) -> Result<SignedMetadata<D, LinkMetadata>>
+  where
+      D: DataInterchange,
+  {
+      SignedMetadata::new(&self.build()?, private_key)
+  }
+}
+
+/// link metadata
+#[derive(Debug, Clone, PartialEq)]
+pub struct LinkMetadata {
```

Comment:
Possibly a silly question, but do we need both `LinkMetadata` here and `Link` in `src/models/link/mod.rs`? Can we roll them into one?

Reply:
Hmm, I'm not too sure. `LinkMetadata` and `Link` were originally from 2 different locations, actually. 
- `Link` was originally one of the TUF `Role` enum (with the other option being `Layout` but commented out because it's unimplemented). 
- `LinkMetadata` was located in the `interchange` cjson folder. It is in a file called `shims.rs` inside the cjson folder and includes more detailed functionalities such as writing links (that we're using in our examples folder)

If we role them into one, would we change the `LinkMetadataBuilder` (what we've been calling for link generation into `LinkBuilder`?